### PR TITLE
[dagster-pipes] arbitrary metadata for PipesClientCompletedInvocation

### PR DESCRIPTION
**Pull request by danielgafni**
## Summary & Motivation

Currently `Pipes` clients only return information and metadata from the external process side and there is no way to add information only available on the orchestration side.

This PR allows passing metadata to `PipesClientCompletedInvocation` invocation. This metadata will be attached to all Pipes materializations (explicit or implicit).  

## How I Tested These Changes

Added tests

## Changelog

Added the ability to pass arbitrary metadata to `PipesClientCompletedInvocation`. This metadata will be attached to all materializations and asset checks stored in this object.

### CODE REVIEW

**Comment by danielgafni** [comment]
* **#24342** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24342?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danielgafni and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by schrockn — CHANGES_REQUESTED** [review]
Can you stack a concrete usage on top of this? That is very helpful for understanding motivation etc

**Comment by schrockn** [comment]
The example is helpful. How does that metadata end up being _consumed_? Is it attached somehow to an event that ends up in the event log?

A _complete_ explanation of the value proposition from the user perspective would be very helpful, and I think would lead to a better name than `client_metadata`.

File: python_modules/dagster/dagster/_core/pipes/client.py [hunk 1]
```diff
@@ -54,8 +54,15 @@ class PipesClientCompletedInvocation:
     def __init__(
         self,
         session: PipesSession,
+        client_metadata: Optional[Dict[str, Any]] = None,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/pipes/client.py:57]: We should have a docstring here.

**Review by schrockn — DISMISSED** [review]
Comment re: docs and motivating example.

**Comment by danielgafni** [comment]
> How does that metadata end up being consumed

That's up to the user. Typically one would store it with `context.add_output_metadata`. The reason we can't do this automatically is because there may be multiple outputs (e.g. `@multi_asset`) and we can't know which one to use.

We can look into logging it automatically if there is only 1 output in the current `OpExecutionContext`. Otherwise log a warning and ask the user to set the output name per each metadata value or something like this. Not sure if we want to go this far. Any thoughts @schrockn ?

File: python_modules/dagster/dagster/_core/pipes/client.py (hunk 1 above)
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/client.py]: done

**Comment by smackesey** [comment]
>We can look into logging it automatically if there is only 1 output in the current OpExecutionContext. Otherwise log a warning and ask the user to set the output name per each metadata value or something like this. Not sure if we want to go this far.

I think we should consider whether we want to automatically attach this metadata to all the materializations and check results returned by `PipesClientCompletedInvocation.get_results()` (and its other variants). The tricky thing is that there are two _kinds_ of results returned by `get_results`:

- implicitly generated results for when the process expected to materialize an asset completed successfully but the external process didn't explicitly pass a materialization event as a message
- explicitly generated results that are just forwarding the materialization events the user already passed

If there were only implicitly generated results, I think the answer is an obvious yes. But the possibility of explicit results complicates things. It's a little bit weird to tack stuff onto objects the user explicitly constructed. But I would still say we should do it, since they always have the escape hatch of not using a client, and I don't think we want different behavior for implicit/explicit results.

**Comment by danielgafni** [comment]
@smackesey finally got back to this, could you take a look again please?

File: python_modules/dagster/dagster/_core/pipes/client.py [hunk 2]
```diff
@@ -52,11 +52,27 @@ def run(
 
 
 class PipesClientCompletedInvocation:
+    """A wrapper for the results of a pipes client invocation, typically returned from `PipesClient.run`.
+
+    Args:
+        session (PipesSession): The Pipes session that was used to run the external process.
+        metadata (Optional[Mapping[str, Any]]): Arbitrary metadata attached to the invocation,
+            such as an external job_id or other information available on the client side. Will be injected
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/client.py:60]: "Arbitrary metadata that will be attached to all asset results (e.g. materializations) generated by the invocation. Useful for attaching information to asset results that is available via the external process launch API but not in the external process itself (e.g. a `job_id` param returned by the launch API call)."

File: python_modules/dagster/dagster/_core/pipes/client.py [hunk 3]
```diff
@@ -52,11 +52,27 @@ def run(
...
+            into materialization results from the remote process.
+    """
+
     def __init__(
         self,
         session: PipesSession,
+        metadata: Optional[Mapping[str, Any]] = None,
     ):
         self._session = session
+        self._metadata = metadata or {}
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/client.py:70]: Let's accept a `RawMetadataMapping` and call `normalize_metadata` on it.

File: python_modules/dagster/dagster/_core/pipes/client.py [hunk 4]
```diff
@@ -52,11 +52,27 @@ def run(
...
+
+    @property
+    def metadata(self) -> Mapping[str, Any]:
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/client.py:73]: Should return `Mapping[str, MetadataValue]`

File: python_modules/dagster/dagster/_core/pipes/context.py [hunk 5]
```diff
@@ -417,28 +417,40 @@ def get_results(
...
-        implicit = (
+        implicit = [
             MaterializeResult(asset_key=key)
             for key in self.context.selected_asset_keys
             if key not in reported_keys
-        )
+        ]
+
+        if metadata:
+            implicit = self._inject_metadata_to_materialization_results(implicit, metadata)
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/context.py:453]: We can just call this once on the combined set of results.

File: python_modules/dagster/dagster/_core/pipes/context.py [hunk 6]
```diff
@@ -475,6 +487,19 @@ def report_launched(self, launched_payload: PipesLaunchedData):
...
+        self, results: Sequence[PipesExecutionResult], metadata: Mapping[str, Any]
+    ) -> Sequence[PipesExecutionResult]:
+        results_with_metadata = []
+        for i, result in enumerate(results):
+            if isinstance(result, MaterializeResult):
+                # .with_metadata returns a new instance
+                results_with_metadata.append(result.with_metadata(metadata))
+            else:
+                # do not modify asset checks
+                results_with_metadata.append(result)
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/context.py:500]: Why are we not modifying asset checks? Is it not useful to have e.g. the same `job_id` on them?

File: python_modules/dagster/dagster/_core/pipes/context.py [hunk 7]
```diff
@@ -475,6 +487,19 @@ def report_launched(self, launched_payload: PipesLaunchedData):
         """
         self.message_handler.on_launched(launched_payload)
 
+    def _inject_metadata_to_materialization_results(
+        self, results: Sequence[PipesExecutionResult], metadata: Mapping[str, Any]
+    ) -> Sequence[PipesExecutionResult]:
+        results_with_metadata = []
+        for i, result in enumerate(results):
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/pipes/context.py:494]: nit: unnecessary `enumerate()`

File: python_modules/dagster/dagster_tests/execution_tests/pipes_tests/in_process_client.py [hunk 8]
```diff
@@ -105,6 +105,7 @@ def run(
         context: Union[OpExecutionContext, AssetExecutionContext],
         fn: Callable[[PipesContext], None],
         extras: Optional[PipesExtras] = None,
+        metadata: Optional[Mapping[str, Any]] = None,  # metadata to attach to all materializations
```
Comment (smackesey) [inline python_modules/dagster/dagster_tests/execution_tests/pipes_tests/in_process_client.py:108]: `RawMetadataMapping`

**Review by smackesey — COMMENTED** [review]
Pls address comments

File: python_modules/dagster/dagster/_core/pipes/client.py (hunk 2 above)
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/client.py:60]: I'm always impressed by your docstrings!

File: python_modules/dagster/dagster/_core/pipes/context.py (hunk 6 above)
Comment (danielgafni) [reply python_modules/dagster/dagster/_core/pipes/context.py:500]: Actually no idea why I decided to skip them. It makes sense to always inject this metadata. Probably brain lag

**Comment by danielgafni** [comment]
Just found out that `AssetCheckResult` already has this `with_metadat` method, but it replaces metadata instead of only injecting new keys. 

image

So I'll just add a separate helper function for injecting metadata (specific to `PipesSession`) to avoid confusion.

**Review by smackesey — APPROVED** [review]
one comment but then good to go

**Comment by danielgafni** [comment]
### Merge activity

* **Nov 8, 6:35 PM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/24342).

**Pull request merged**